lib/src/tls.rs

```diff
@@ -0,0 +1,1038 @@
+            .parse_x509()
+            .map_err(|err| GenericCertificateResolverError::DerParseError(err.to_string()))?;
+
+        // We need to know if the new certificate can replace an already existent one.
+        let new_names = match self.get_names_override(fingerprint) {
+            Some(names) => names,
+            None => self.certificate_names(&parsed_certificate_and_key.certificate)?,
+        };
+
+        let expiration = self
+            .get_expiration_override(fingerprint)
+            .unwrap_or_else(|| x509.validity().not_after.timestamp());
+
+        let fingerprints = self.find_certificates_by_names(&new_names)?;
+        let mut certificates = HashMap::new();
+        for fingerprint in &fingerprints {
+            if let Some(certificate_and_key) = self.get_certificate(fingerprint) {
+                certificates.insert(fingerprint, certificate_and_key);
+            }
+        }
+
+        let mut should_insert = false;
+        let mut certificates_to_remove = HashMap::new();
+        let mut certificates_names = HashSet::new();
+        for (fingerprint, certificate_and_key) in certificates {
+            let certificate = certificate_and_key
+                .certificate
+                .parse_x509()
+                .map_err(|err| GenericCertificateResolverError::DerParseError(err.to_string()))?;
+
+            // could not failed as this operation has been already done when loading
+            // the certificategit
+            let certificate_names = match self.get_names_override(&fingerprint) {
+                Some(names) => names,
+                None => self.certificate_names(&parsed_certificate_and_key.certificate)?,
+            };
+
+            let certificate_expiration = self
+                .get_expiration_override(&fingerprint)
+                .unwrap_or_else(|| certificate.validity().not_after.timestamp());
+
+            let extra_names = certificate_names
+                .difference(&new_names)
+                .collect::<HashSet<_>>();
+
+            // if the certificate has at least the same name or less and the expiration date
+            // is closer than the new one. We could remove it and allow the new insertion.
+            if extra_names.is_empty() && certificate_expiration < expiration {
+                certificates_to_remove.insert(fingerprint.to_owned(), certificate_names.to_owned());
+                should_insert = true;
+            }
+
+            // We keep a track of all name of certificates that match our query to
+            // check, if the new certificate provide an extra domain which is not
+            // already exposed
+            for name in certificate_names {
+                certificates_names.insert(name);
+            }
+        }
+
+        // In the case, where we do not insert the certificate, because there is
+        // no additional value. We have to check, if it provide an extra domain
+        // name not registered yet.
```

Comment:
```suggestion
        // In the case where we do not insert the certificate because there is
        // no additional value, we have to check wether it provides an extra, not
        // yet registered domain name.
```